pkg/controllers/registryclient/internal/orphandeleter/orphan_deleter.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package orphandeleter
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/log"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	connectivityclientset "github.com/vmware-tanzu/cross-cluster-connectivity/pkg/generated/clientset/versioned"
+	connectivitylisters "github.com/vmware-tanzu/cross-cluster-connectivity/pkg/generated/listers/connectivity/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type OrphanDeleter struct {
+	serviceRecordLister  connectivitylisters.ServiceRecordLister
+	connClientSet        connectivityclientset.Interface
+	namespace            string
+	deleteOrphanTimer    *time.Timer
+	deleteOrphanDelay    time.Duration
+	remoteServiceRecords map[types.NamespacedName]struct{}
+
+	mutex sync.RWMutex
+}
+
+func NewOrphanDeleter(
+	serviceRecordLister connectivitylisters.ServiceRecordLister,
+	connClientSet connectivityclientset.Interface,
+	namespace string,
+	deleteOrphanDelay time.Duration,
+) *OrphanDeleter {
+	return &OrphanDeleter{
+		serviceRecordLister:  serviceRecordLister,
+		connClientSet:        connClientSet,
+		namespace:            namespace,
+		remoteServiceRecords: map[types.NamespacedName]struct{}{},
+		deleteOrphanDelay:    deleteOrphanDelay,
+	}
+}
+
+func (o *OrphanDeleter) Reset() {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+	if o.deleteOrphanTimer != nil {
+		o.deleteOrphanTimer.Stop()
+		o.deleteOrphanTimer = nil
+	}
+	o.remoteServiceRecords = map[types.NamespacedName]struct{}{}
+}
+
+func (o *OrphanDeleter) AddRemoteServiceRecord(namespacedName types.NamespacedName) {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	if o.deleteOrphanTimer == nil {
+		o.deleteOrphanTimer = time.AfterFunc(o.deleteOrphanDelay, o.deleteOrphans)
+	}
+
+	o.remoteServiceRecords[namespacedName] = struct{}{}
+}
+
+func (o *OrphanDeleter) copyOfRemoteServiceRecords() map[types.NamespacedName]struct{} {
+	remoteServiceRecords := map[types.NamespacedName]struct{}{}
+	o.mutex.RLock()
+	defer o.mutex.RUnlock()
+	for k, v := range o.remoteServiceRecords {
+		remoteServiceRecords[k] = v
+	}
+	return remoteServiceRecords
+}
+
+func (o *OrphanDeleter) deleteOrphans() {
+	log.Infof("Cleaning up orphaned service records")
+
+	importedServiceRecords, err := o.listServiceRecords()
+	if err != nil {
+		log.Errorf("error listing service records: %s", err)
+	}
+
+	remoteServiceRecords := o.copyOfRemoteServiceRecords()
```

Comment:
Curious. Why do you need to make a copy of the map here? Is it because you wanted to reduce the amount of time necessary to keep the lock? i.e `deleteOrphans` is a slow process so you make a copy?

Reply:
Yes, that was the intention. Originally the lock was being held while several kubeapi calls were being made. This change means the lock is only held momentarily.

Reply:
Cool makes sense.